ppcs/ppcTODO-signature-named-parameters.md

```diff
@@ -0,0 +1,153 @@
+# Named Parameters in Signatures
+
+## Preamble
+
+    Author:  Paul Evans <PEVANS>
+    Sponsor:
+    ID:      TODO
+    Status:  Exploratory
+
+## Abstract
+
+Adds the ability for subroutine signatures to process named arguments in the form of name-value pairs passed by the caller, in a fashion familiar to existing uses of assignment into a hash.
+
+## Motivation
+
+Perl 5.20 added "subroutine signatures", native syntax for more succinctly expressing the common patterns of processing arguments inbound into a subroutine by unpacking the `@_` array. Rather than writing common styles of code directly, a more compact notation, in a style immediately familiar to users of many other languages, allows Perl to create the required behaviour from that specification.
+
+```perl
+sub f ($x, $y, $z = 123) { ... }
+
+# instead of
+sub f {
+    die "Too many arguments" if @_ > 3;
+    die "Too few arguments" if @_ < 2;
+    my ($x, $y, $z) = @_;
+    $z = 123 if @_ < 3;
+    ...
+}
+```
+
+One common form of argument processing involves passing an even-sized list of key/value pairs and assigning that into a hash within the subroutine, so that specifically named parameters can be extracted from it. There is currently no support from subroutine signature syntax to assist authors in providing such behaviours.
+
+## Rationale
+
+(explain why the (following) proposed solution will solve it)
+
+## Specification
+
+A new kind of element may be present in a subroutine signature, which consumes a named argument from the caller. These elements are written with a leading colon prefix (`:$name`), indicating that it is named rather than positional. The name of each parameter is implied by the name of the lexical into which it is assigned, minus the leading `$` sigil.
+
+Each element provides a new lexical variable that is visible during the body of the function, in the same manner as positional ones.
+
+The value of a named parameter is taken from the argument values passed by the caller, in a manner familiar to existing uses of hash assignment. The caller should pass an even-sized name-value pair list. The values corresponding to names of parameters will be assigned into the variables. The order in which the values are passed by the caller is not significant.
+
+Furthemore, all of the new behaviour is performed within the body of the invoked subroutine entirely by inspecting the values of the arguments that were passed. The subroutine is not made aware of how those values came to be passed in - whether from literal name-value syntax, a hash or array variable expansion, or any other expression yielding such a list of argument name and value pairs.
+
+```perl
+sub make_colour ( :$red, :$green, :$blue ) { ... }
+
+make_colour( red => 1.0, blue => 0.5, green => 0.2 );
+# The body of the function will be invoked with
+#   $red   = 1.0
+#   $green = 0.2
+#   $blue  = 0.5
+```
```

Comment:
It's hard for me to follow, but does this not give you the unpack equivalent of `my %param = @_` when unambiguous?

Reply:
If you're suggesting that this should give `$param = { red => 1.0, green => 0.2, blue => 0.5 }`, that'd make this entire thing PPC pointless: why not just `sub make_colour (%param)` in that case? You can already do that now, but it doesn't provide typo protection (`$param{bleu}`) and everything now has to go through that extra hash indirection - more typing involved, closures end up with the entire hash of parameters instead of the individual values, etc.

If the suggestion is that the _callsite_ should be a hashref, that adds extra syntax and I think makes the code less clear: why would named parameters need to be a hashref instead of a list? It's not going to end up in a hashref in the sub body.

Reply:
Then this whole thing is pointless. When anyone is passing in named parameters with `=>`, shoving them into a litany of scalars is forcing them in the baby Perl tier; I'm willing to bet 99% of the time the author wants a HASH ref.

Reply:
The common presence of this idiom across CPAN and constant requests for the feature would suggest it is in fact not pointless.

Reply:
> Then this whole PPC is pointless. When anyone is passing in named parameters with `=>`, shoving them into a litany of scalars is forcing them in the baby Perl tier; I'm willing to bet 99% of the time the author wants a HASH ref to keep things organized in the `sub`. It's a bait-n-switch. You're guaranteeing anyone who has any sense of idiomatic Perl will _not_ use this. If there are more advanced cases where this actually pays off, please show them. Because that RBG example is extremely lame.

I partially agree with you, especially when using slurpy hash, which may contain same keys as signature parameters.
Personally I think named parameter call syntax should not collide with any other syntax, which leads to two available solution:
- prefix named parameters with token
- prefix slurpy parameters with token (similar to `--` recognized by many command line tools)

Just returning to my earlier proposal:
```
make_colour ($r, $b, := $green => ...);
```

(which allows passing multiple lists as well has possible benefits beyond passing just variables)

Reply:
> Then this whole PPC is pointless. When anyone is passing in named parameters with `=>`, shoving them into a litany of scalars is forcing them in the baby Perl tier; I'm willing to bet 99% of the time the author wants a HASH ref to keep things organized in the `sub`. It's a bait-n-switch. You're guaranteeing anyone who has any sense of idiomatic Perl will *not* use this. If there are more advanced cases where this actually pays off, please show them. Because that RBG example is extremely lame.

I gave you specific examples already, though?

- compiletime typo protection
- not having to keep an entire hash of unrelated things around in closures

If you don't use scalars anywhere and just keep everything in hashrefs, that's slightly odd, but your choice - extracting out to scalars is very much idiomatic Perl, though. If you've never seen `my ($x, $y) = @_` or `my $x = delete $args{x} // 'default';` then you're in a very different Perl world, and perhaps this PPC just isn't relevant to you.

Reply:
> Then this whole PPC is pointless. When anyone is passing in named parameters with `=>`, shoving them into a litany of scalars is forcing them in the baby Perl tier; 

I don't think you understand the proposal here.

Reply:
> I don't think you understand the proposal here.

suggesting compromise: hide this behind feature (on call side). When feature is disabled, "forced named" parameters became positional. Then it's up to user to decide.

Reply:
Yes - as @Grinnz points out, this sort of pattern is already extremely common.

It is not the point of this PPC to try to invent some fundamentally new behaviour or feature. It simply attempts to gather up an already-prevalent behaviour and formalise it into a neater way of expression. You can take justabout any CPAN module and likely find somewhere that some code unpacks its arguments into a hash, just to extract out a few fixed named keys into variables. Any of those cases could be rewritten with this feature, gaining the safety of typo protection and shorter, neater code that makes the intent easier to see because of less low-level machinery noise.

Reply:
For t

> > Then this whole PPC is pointless. When anyone is passing in named parameters with `=>`, shoving them into a litany of scalars is forcing them in the baby Perl tier;
> 
> I don't think you understand the proposal here.

Certainly possible, but show some real examples before and after where this would be useful. To most Perl developers this looks like more unnecessary complication, particularly when such an established formalism as presented in Getopt::Long is more than well suited to describe complex named parameters that do not depend on position. It also supports through various means shoving it into single scalars or into a HASH reference. And if you thought it through a bit, you could even make the resulting container class with builtin accessors.

Regarding Getopt::Long's established patterns, this would be immediately be useful for writing modulinos that need to deal with both commandline arguments and the differentiation of `caller` context.

I'd like to see some _common_ patterns, including standard idiomatic unpacking and how this would make it better.

Additinally, this seems like it could wreak havoc on any work that's been done so far on multidispatch - which unlike this PPC - seems to be actually useful to the regular Perl developer.

In summary - we need more examples so it can be evaluated by regular folks. The language would develop a lot faster and more in service to the daily Perl jocky/involuntary code maintainer if there was some separation between language design and implementation. This is just another of example of how inefficient this process is. The person who does most of the "new" development should not also be the one proposing "new" ideas that sure have been developed in a vacuum.

Reply:
Perhaps you should read the `EXAMPLES` section of this very document, to find some examples to demonstrate it.

Reply:
@oodler577 For reference, here's a query to show some of the standard idiomatic examples from CPAN:

https://grep.metacpan.org/search?q=my+%5CS%2B+%3D+delete+%5C%24args%5C%7B&qd=&qft=&qifl=inc%2FModule%2FInstall%2FMakefile.pm

Currently lists over 700 distributions, give it some time to work through the rest of the search index and I'd expect that number to grow. This would seem to qualify as a _common_ pattern.

In each case, multiple lines condense into a single signature without the repetition of the name and `delete $args{...}`. That's directly useful: as a regular user, I'd be happy to use it. I know this because I'm _already_ using it, in fact, via the prototype [`Sublike::Extended`](https://metacpan.org/pod/Sublike::Extended) module mentioned in this very document. It is quite nice, and has already cut down on some finger and eyeball wear and tear, similar to the way `x // y` simplified `defined(x) ? x : y`.

As for implementing `Getopt::Long` in the signature parser, that seems like a wonderfully strange suggestion to counter "unnecessary complication", albeit not a _useful_ suggestion absent any examples or specifics. Lamentations about efficiency of process or separation of "designer" and "implementor" best directed to p5p or Facebook, though.

Reply:
Hashes have overhead on access and no inherent advantage with storage. A hash ref was never designed to "organize" or namespace variables or pass arguments to a sub. It's a workaround and it comes at a cost (you're using a datastructure to workaround language semantics).

That said, this isn't without drawback. A function that takes a hash ref will have flatten the hashref it takes to pass the arguments to the function requiring named params.

```perl
sub f1 ( $args )  {
  f2( %$args );
}

sub f2 ( $x, $y, $z ) {
}
```

It was said earlier,

> The whole point as the document states, is that the **callsites for these functions do not need to be changed or know about it.**

That's only kinda-true. An advanced Perl user will only pass around hashrefs because

* it's faster than passing around scalars
* it can save the listify-hashify dance on each call.

This convention will push people into writing subs with expect lists which will make the above convention, the status quo, awkward. The real question here is whether or not the added value is worth it.

Reply:
I think if we pursue destructuring assignments as mentioned in a different comment thread then the hashref vs flat list question is less problematic, bc you can unpack the kwargs hashref in one line w the same protections that signatures would provide.
Would make it one step harder for tooling like perlnavigator to help with, but I think that can be dealt with without too much trouble.

I'm a little averse to adding in a syntax to unpack kwargs from a hashref to signatures, so I would also say flat hashes ftw.

Reply:
Advanced Perl users might use hashref parameters in some places, and that's not something that's going away, but I don't think the hashref/hash performance difference is that clear-cut.

Hashrefs can be a bit _slower_ when you're constructing a new/modified set of parameters - which is pretty common, compared to subs/methods that pass exactly the same parameters on to something else? It's also all too easy for something you call to remove/change something in the hashref that affects the rest of your sub:

```perl
 sub f ($args) {
  say $args->{something};
  g($args);
  say $args->{something}; # it was there just a moment ago :/
 }
 sub g ($args) {
  my $value = delete $args->{something};
 }
```

Unless you're only calling your own code, or what you're calling documents strict guarantees on not tampering with the hashref, I find that an "only-ever-hashref" API policy ends up causing more harm than good :smile: 

A simple benchmark for context here:

```
$ perl -e'
 use 5.040;
 use Benchmark qw(cmpthese :hireswallclock);
 use Sublike::Extended;
 extended sub manual (%args) {
  my $x = delete $args{x} or die "need x";
  my $y = delete $args{y} or die "need y";
  my $z = delete $args{z} or die "need z";
  die "leftover parameters" if %args;
  $x + $y + $z
 }
 extended sub named (:$x, :$y, :$z) { $x + $y + $z }
 extended sub hashref ($args) { $args->{x} + $args->{y} + $args->{z} }
 extended sub hash (%args) { $args{x} + $args{y} + $args{z} }
 cmpthese -2, {
  manual => sub { die unless 12 == manual(x => 3, y => 4, z => 5) },
  named => sub { die unless 12 == named(x => 3, y => 4, z => 5) },
  hashref => sub { die unless 12 == hashref({ x=> 3, y => 4, z => 5 }) },
  hash => sub { die unless 12 == hash(x => 3, y => 4, z => 5) }
 }
'
             Rate  manual   named hashref    hash
manual  4128545/s      --     -1%    -14%    -26%
named   4150703/s      1%      --    -14%    -26%
hashref 4812082/s     17%     16%      --    -14%
hash    5614095/s     36%     35%     17%      --
```

Reply:
> Perhaps you should read the `EXAMPLES` section of this very document, to find some examples to demonstrate it.

Thank you. The examples were the ones I saw from the PR, but more clear in https://github.com/leonerd/Perl-PPCs/blob/signature-named-params/ppcs/ppcTODO-signature-named-parameters.md.

They are still all just straight assignments for a hash key-value to a scalar. Your examples are not idiomatic and I believe contrived or cherry-picked.

It seems this is useful for a very limited sub-case where you have just a few named parameters and thus assigning key values to singular scalars is useful; like I said above, automatically replacing a litany of `my $foo = $arg{foo}` is not useful. Yeah you remove many lines of boiler plate, but that kind of pattern is not idiomatic at all anyway; hence the idiom of 

```
...
sub foo {
  my ($self, %args) = @_;
  # put some validation here if you want
}

...

my $whatever = $classic_obj_Foo->foo(arg1 => "val1", arg2 => \@val2, ... argN => $valN);
```

One non-idiomatic thing I do because I often don't have the wherewithal to look up how to do it is taking `@_` and putting it directly into a HASH ref.

What would be actually useful,

```
sub say_foo (some signature defining DSL thing ) {
  my ($self, $args_ref) = @_;
  say $args_ref->{foo};
  ...
}

$classic_obj_Foo->say_foo(foo => "hello, world!");
```

Or if you really wanted to win some internet points, add a getter:

```
sub say_foo (some signature defining DSL thing ) {
  my ($self, $args_ref) = @_;
  say $args_ref->foo;
  ...
}

$classic_obj_Foo->say_foo(foo => "hello, world!");
```

And yes, @tm604, considering the benefits of what Getopt::Long does to define a "standard" (that most of us probably know pretty well, I assume) would be great. It's familiar and battle tested. It is also supremely helpful in the use case of a modulino, as I mentioned above. Having subroutines that service a modulino that cares about `@ARGS` is very useful for testing, it it also avoids a translation layer for `@ARGS` if you're writing unit tests for said modulino.

Can you address the implications here for Corinna? I have to believe this thing that appears unhelpful in this isolated "normal Perl code" presentation is going to yield benefits down the line there? It has been pointed out elsewhere that this is more useful for the basis of "type" validation than anything else.

Reply:
> performance

There is a time and place for performance considerations, this is about language semantics and idiomatic quality of life improvements. So this is not one of those times.

Reply:
> Your examples are not idiomatic and I believe contrived or cherry-picked.

Just as a data point here, I performed the following on my current job's work repo:
```
ag '\$\w+ = \$?\$args' | wc -l
20441
```
I've learned very quickly the last couple of months that just because _I_ don't use a idiom doesn't mean that idiom isn't popular.

Basically I'm saying that if this is cherry-picked there is evidence that it's a pretty damn big cherry to contrive.

Reply:
So people who do that are going to figure out how to define a complex signature DSL statement, people who otherwise were okay with straight assignments? I mean the medicine is way worse than the cure of any potential issue. I stand of the opinion that this proposal does not stand on its own. I am looking for some situations where it does. Is it just an incomplete PPC? That's fair if it is. Just make it complete and maybe the cost of this added complexity will be more apparent to me.

Reply:
named parameters appeal to me but adding the :$param syntax in the signature section feels heavy on my mind when considering how feature-rich our subroutines already are. Might I offer a possibility to help you avoid the : and distinguish it from the other style arguments? Perhaps an optional block that contains named parameters?
```perl
sub make_color ( $arg1, @slurpy) < $red = "red",  $blue = "blue"> {
     say "sub's body";
}
```
they < > wrappers may not be the best choice.. but the point is that a different/separate wrapper provides a context for less contrived syntax in the named var assignments and/or confusion about how params work in perl to newcomers

Reply:
@haxmeister The proposal is not that far off from other languages. Raku uses a similar syntax. The colon prefix is similar to the convention of using a colon for keyword symbols in Common Lisp (though this is at the call site).

Also, all languages I am familiar with that have named parameters allow for a mix of positional and named parameters in a single parameter list in their signatures, so that shouldn't be novel.

Reply:
> @haxmeister The proposal is not that far off from other languages. Raku uses a similar syntax. The colon prefix is similar to the convention of using a colon for keyword symbols in Common Lisp (though this is at the call site).
> 
> Also, all languages I am familiar with that have named parameters allow for a mix of positional and named parameters in a single parameter list in their signatures, so that shouldn't be novel.

Just to point out, most of languages doesn't have sigils and treat list variables as scalars, so not providing sigil in call syntax is designed to lose information.

Reply:
The syntax by itself is not the reason that I found this to be heavy on my mind. Rather it it the syntax in light of how it is used in the current class feature. I have already associated the  colon as an attribute syntax that follows a convention where the colon is followed by the name of a behavior that is applied to the preceding var. This proposed syntax will apply the colon directly to the var.. it is not the same thing and my mind immediately rejected using it this way after having written much code using the new class syntax.

Reply:
@haxmeister, yeah, I can see that. I like the extensibility of the KIM syntax that was mentioned before in the design of the `class` feature.

But choosing a single ASCII character not used elsewhere in the Perl syntax is also difficult. Named parameters are probably going to be a commonly used feature so shorter is better.

Perhaps looking at how Kavorka does it <https://metacpan.org/dist/Kavorka/view/lib/Kavorka/Manual/Signatures.pod#Named-parameters>, <https://metacpan.org/dist/Kavorka/view/lib/Kavorka/Manual/Signatures.pod#Long-name-parameters> which is based on the form used in Raku, might help.

If the approach that @leonerd has of having attributes in the parameter list goes through, I think using a longer form with a `:name()` parameter attribute that repeats the same name as the variable could be a pattern to make the syntax stand out. I typically write out anything with many parameters and additional default values so that each parameter is on a line by itself.